submariner/OVN-Interconnect.md

```diff
@@ -0,0 +1,244 @@
+endpoints.
+
+#### SubmarinerGWRoute CR Controller
+
+This controller will be responsible for programming the OVN cluster router, it will react only the active g/w node. When a SubmarinerGWRoute
+CR is created or modified the controller shall create or update a routing policy in OVN cluster router with a priority of 20000, and it should
+redirect any traffic destined to remote CIDR to the ovn-k8s-mp0 interface IP.
+
+```bash
+_uuid               : 0459f009-3603-47ac-8ee7-9d958540ed31
+bfd                 : []
+external_ids        : {}
+ip_prefix           : "10.132.0.0/16"
+nexthop             : "10.1.1.2"
+options             : {}
+output_port         : []
+policy              : []
+route_table         : ""
+```
+
+It also programs a route in the ovn-cluster-router, to route the traffic coming from other zones destined to remote cluster IP range via the
+ovn-k8s-mp0 interface IP.
+
+```bash
+_uuid               : d55185d8-3732-45c1-ae90-4a7f8cd191f7
+bfd                 : []
+external_ids        : {}
+ip_prefix           : "10.132.0.0/16"
+nexthop             : "10.1.1.2"
+options             : {}
+output_port         : []
+policy              : []
+route_table         : ""
+```
+
+#### SubmarinerNonGWRoute Controller
+
+This controller will run in every route agent pod. This controller connects to the OVN DB. When a SubmarinerNonGWRoute CR is created
+in non-g/w node,  it updates the ovn-cluster-route with a router policy using a priority of 20000 to send the traffic to
+the remote cluster via next hop mentioned, which is the transit switch IP to the g/w node. Before adding the route it checks if
+a route exists, if so it skips adding the route again. This is required to prevent duplicate update since there can be more than
+one node in each zone and hence more than one RouteAgent.
+
+```bash
+_uuid               : 22db3005-64c5-4e32-aeb0-642423c30742
+action              : reroute
+external_ids        : {}
+match               : "ip4.dst==10.132.0.0/16"
+nexthop             : []
+nexthops            : ["169.254.0.1"]
+options             : {"external_ids:{submariner"="true}"}
+priority            : 20000
+```
+
+### Backward Compatibility
+
+With this architecture, we are removing the logical switches and routers that Submariner created in the existing implementation.
+During migration, we will need to delete these components and the routes that network-plugin-syncer installed. After that,
+the new controllers shall install the updated routes. The operator needs to be updated as well, not to create a
+network-plugin-syncer pods and remove any existing deployments.
+
+### Gateway FailOver
+
+If there are two gateway nodes , the passive one will work like a non-gateway node. It will not be responsible for creating the CRs.
+In the case of gateway fail-over all the current SubmarinerGWRoute and SubmarinerNonGWRoute will be deleted by the route agent in
```

Comment:
I don't think we can always ensure that all CRs will be deleted by the route agent in the node that is transitioning to non-gateway node, I assume that it is relevant for graceful switchover.

What value will be used as CR key? is it CR name?

Reply:
it will be the name and will same as endpoint name. We can always add a clean up logic when a node transition to g/w node. It is more a implementation detail.